libs/ml/include/ml/model/model.hpp

```diff
@@ -259,11 +298,33 @@ void Model<TensorType>::Predict(TensorType &input, TensorType &output)
 }
 
 template <typename TensorType>
-typename Model<TensorType>::DataType Model<TensorType>::Evaluate(
-    std::vector<MetricType> const &metrics)
+typename Model<TensorType>::DataVectorType Model<TensorType>::Evaluate(
+    dataloaders::DataLoaderMode dl_mode, SizeType batch_size)
 {
-  FETCH_UNUSED(metrics);
-  return loss_;
+  if (!compiled_)
+  {
+    throw ml::exceptions::InvalidMode("must compile model before evaluating");
+  }
+
+  dataloader_ptr_->SetMode(dl_mode);
+  bool is_done_set;
+  if (batch_size == 0)
+  {
+    batch_size = dataloader_ptr_->Size();
+  }
+  auto evaluate_pair = dataloader_ptr_->PrepareBatch(batch_size, is_done_set);
+
+  this->graph_ptr_->SetInput(label_, evaluate_pair.first);
+  this->graph_ptr_->SetInput(input_, evaluate_pair.second.at(0));
+
+  DataVectorType ret;
+  ret.emplace_back(*(this->graph_ptr_->Evaluate(error_).cbegin()));
+
+  for (auto const &met : metrics_)
+  {
+    ret.emplace_back(*(this->graph_ptr_->Evaluate(met).cbegin()));
```

Comment:
have you confirmed (via debugging or benchmarking) that the caching is doing it's job as we are hoping here - i.e. that these evaluate calls are very cheap. if not, please step through in debugging at least and just check it works how we expect.

Reply:
Yep, checked in the debugger and cache status is CACHE_VALID